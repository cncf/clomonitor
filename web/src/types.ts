export interface Organization {
  name: string;
  display_name?: string;
  description?: string;
  home_url: string;
  logo_url?: string;
}

export interface BaseProject {
  id: string;
  name: string;
  display_name?: string;
  description?: string;
  home_url: string;
  logo_url?: string;
  devstats_url?: string;
  maturity_id: Maturity;
  category_id: Category;
  score: Score;
  updated_at: number;
}

export interface Project extends BaseProject {
  repositories: BaseRepository[];
  organization: {
    name: string;
  };
}

export interface ProjectDetail extends BaseProject {
  repositories: Repository[];
}

export interface BaseRepository {
  name: string;
  url: string;
  kind: RepositoryKind;
}

export interface Repository extends BaseRepository {
  digest: string;
  repository_id: string;
  score: Score;
  reports: Report[];
}

export interface Report {
  data: CoreReport | any;
  linter_id: LinterId;
  report_id: string;
  updated_at: number;
}

export interface CoreReport {
  // ScoreType
  [key: string]: {
    [key: string]: ReportCheck;
  };
}

export interface ReportCheck {
  passed: boolean;
  value?: string;
  url?: string;
}

export type Score = {
  [key in ScoreType]: number;
} & { score_kind: ScoreKind };

export interface FiltersSection {
  name: string;
  title: string;
  filters: Filter[];
}

export interface Filter {
  name: string | number;
  label: string;
  legend?: string;
  decorator?: JSX.Element;
}

export interface Issue {
  level: number;
  description: string;
}

export interface Prefs {
  search: { limit: number; sort: { by: SortBy; direction: SortDirection } };
  theme: {
    effective: string;
  };
}

export interface ReportOptionData {
  icon: JSX.Element;
  name: string;
  legend: JSX.Element;
}

export enum Maturity {
  Graduated = 0,
  Incubating,
  Sandbox,
}

export enum Category {
  'App definition' = 0,
  Observability,
  Orchestration,
  Platform,
  Provisioning,
  Runtime,
  Serverless,
}

export enum Rating {
  A = 'a',
  B = 'b',
  C = 'c',
  D = 'd',
}

export enum FilterKind {
  Maturity = 'maturity',
  Category = 'category',
  Rating = 'rating',
}

export enum LinterId {
  core = 0,
}

export enum ScoreType {
  BestPractices = 'best_practices',
  Documentation = 'documentation',
  Global = 'global',
  Legal = 'legal',
  License = 'license',
  Security = 'security',
}

export enum SortDirection {
  ASC = 'asc',
  DESC = 'desc',
}

export enum SortBy {
  Name = 'name',
  Score = 'score',
}

export enum RepositoryKind {
  Primary = 'primary',
  Secondary = 'secondary',
}

export enum ScoreKind {
  Primary = 'primary',
  Secondary = 'secondary',
}

export enum ReportOption {
  Adopters = 'adopters',
  ApprovedLicense = 'approved',
  ArtifactHubBadge = 'artifacthub_badge',
  Changelog = 'changelog',
  CodeOfConduct = 'code_of_conduct',
  CommunityMeeting = 'community_meeting',
  Contributing = 'contributing',
  DCO = 'dco',
  Governance = 'governance',
  LicenseScanning = 'scanning',
  Maintainers = 'maintainers',
  OpenSSFBadge = 'openssf_badge',
  Readme = 'readme',
  RecentRelease = 'recent_release',
  Roadmap = 'roadmap',
  SecurityPolicy = 'security_policy',
  SlackPresence = 'slack_presence',
  SPDX = 'spdx_id',
  TrademarkDisclaimer = 'trademark_disclaimer',
  Website = 'website',
}

export interface SearchFiltersURL extends BasicQuery {
  pageNumber: number;
}

export interface BasicQuery {
  text?: string;
  filters?: {
    [key: string]: (string | number)[];
  };
}

export interface SearchQuery extends BasicQuery {
  limit: number;
  offset: number;
  sort_by: SortBy;
  sort_direction: SortDirection;
}

export interface SearchData {
  limit: number;
  offset: number;
  sort_by: string;
  sort_direction: string;
  text?: string;
  category?: number[];
  maturity?: number[];
  rating?: number[];
}

export interface Error {
  kind: ErrorKind;
  message?: string;
}

export enum ErrorKind {
  Other,
  NotFound,
}

export type ReportOptionInfo = {
  [key in ReportOption]: ReportOptionData;
};

export interface RecommendedTemplate {
  name: string;
  url: string;
}
