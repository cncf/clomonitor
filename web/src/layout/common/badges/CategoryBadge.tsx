import { MdCategory } from 'react-icons/md';

import { Category } from '../../../types';
import styles from './Badge.module.css';

interface Props {
  categoryId: Category;
  className?: string;
}

const CartegoryBadge = (props: Props) => {
  return (
    <div
      data-testid="category-badge"
      className={`badge text-secondary border border-secondary rounded-0 position-relative ${styles.badge} ${props.className}`}
    >
      <div className="d-flex flex-row align-items-center">
        <MdCategory className="me-2" />
        {Category[props.categoryId]}
      </div>
    </div>
  );
};

export default CartegoryBadge;
