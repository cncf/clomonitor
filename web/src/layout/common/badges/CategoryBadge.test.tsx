import { render, screen } from '@testing-library/react';

import { Category } from '../../../types';
import CategoryBadge from './CategoryBadge';

describe('CategoryBadge', () => {
  afterEach(() => {
    jest.resetAllMocks();
  });

  it('creates snapshot', () => {
    const { asFragment } = render(<CategoryBadge categoryId={Category['App definition']} />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('renders category 0', () => {
    render(<CategoryBadge categoryId={Category['App definition']} />);
    expect(screen.getByText('App definition')).toBeInTheDocument();
  });

  it('renders category 1', () => {
    render(<CategoryBadge categoryId={Category.Observability} />);
    expect(screen.getByText('Observability')).toBeInTheDocument();
  });

  it('renders category 2', () => {
    render(<CategoryBadge categoryId={Category.Orchestration} />);
    expect(screen.getByText('Orchestration')).toBeInTheDocument();
  });

  it('renders category 3', () => {
    render(<CategoryBadge categoryId={Category.Platform} />);
    expect(screen.getByText('Platform')).toBeInTheDocument();
  });

  it('renders category 4', () => {
    render(<CategoryBadge categoryId={Category.Provisioning} />);
    expect(screen.getByText('Provisioning')).toBeInTheDocument();
  });

  it('renders category 5', () => {
    render(<CategoryBadge categoryId={Category.Runtime} />);
    expect(screen.getByText('Runtime')).toBeInTheDocument();
  });

  it('renders category 6', () => {
    render(<CategoryBadge categoryId={Category.Serverless} />);
    expect(screen.getByText('Serverless')).toBeInTheDocument();
  });
});
