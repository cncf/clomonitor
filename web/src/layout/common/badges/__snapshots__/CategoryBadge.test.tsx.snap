// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CategoryBadge creates snapshot 1`] = `
<DocumentFragment>
  <div
    class="badge text-secondary border border-secondary rounded-0 position-relative badge undefined"
    data-testid="category-badge"
  >
    <div
      class="d-flex flex-row align-items-center"
    >
      <svg
        class="me-2"
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 24 24"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M0 0h24v24H0z"
          fill="none"
        />
        <path
          d="M12 2l-5.5 9h11z"
        />
        <circle
          cx="17.5"
          cy="17.5"
          r="4.5"
        />
        <path
          d="M3 13.5h8v8H3z"
        />
      </svg>
      App definition
    </div>
  </div>
</DocumentFragment>
`;
